core/src/main/java/brooklyn/event/basic/MapConfigKey.java

```diff
@@ -125,6 +131,25 @@ protected Object applyEntryValueToMap(Entry value, Map target) {
         } else if (k instanceof String) {
             k = subKey((String)k);
         } else {
+            // supplier or other unexpected value
+            if (k instanceof Supplier) {
+                Object mapAtRoot = target.get(this);
+                if (mapAtRoot==null) {
+                    mapAtRoot = new LinkedHashMap();
+                    target.put(this, mapAtRoot);
+                }
+                // TODO above is not thread-safe, and below is assuming synching on map 
+                // is the best way to prevent CME's, which is often but not always true
+                if (mapAtRoot instanceof Map) {
+                    if (mapAtRoot instanceof ConcurrentMap) {
+                        return ((Map)mapAtRoot).put(k, value.getValue());
+                    } else {
+                        synchronized (mapAtRoot) {
```

Comment:
Or the supplier of the map might have done `Collections.synchronizedMap(...)`. I'm ok with leaving this synchornized in here, but think that we should never be relying on it - if synchronization is important, then the map should have been created with that in mind (e.g. `Collections.synchronizedMap(...)` means that you never need to synchronize explicitly except for multi-call actions such as iterating).